File: server/bcephal-sheet/src/main/java/com/moriset/bcephal/sheet/domain/SpreadSheetCell.java
/**
 * 
 */
package com.moriset.bcephal.sheet.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.moriset.bcephal.domain.Persistent;
import com.moriset.bcephal.domain.filters.MeasureFilterItem;
import com.moriset.bcephal.domain.filters.UniverseFilter;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * @author Joseph Wambo
 *
 */
@Entity(name = "SpreadSheetCell")
@Table(name = "BCP_SPREADSHEET_CELL")
@Data
@EqualsAndHashCode(callSuper = false)
public class SpreadSheetCell extends Persistent {

	private static final long serialVersionUID = 1585486346524003012L;

	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "spreadsheet_cell_seq")
	@SequenceGenerator(name = "spreadsheet_cell_seq", sequenceName = "spreadsheet_cell_seq", initialValue = 1, allocationSize = 1)
	private Long id;

	@JsonIgnore
	private Long spreadSheet;

	@Enumerated(EnumType.STRING)
	private SpreadSheetType type;

	private String name;

	private int col;

	private int row;

	private String sheetName;

	private int sheetIndex;

	@ToString.Exclude
	@EqualsAndHashCode.Exclude	
	@ManyToOne
	@JoinColumn(name = "cellMeasure")
	private MeasureFilterItem cellMeasure;

	@ToString.Exclude
	@EqualsAndHashCode.Exclude	
	@ManyToOne
	@JoinColumn(name = "filter")
	private UniverseFilter filter;

	private boolean refreshWhenEdit;

	public SpreadSheetCell() {
		this.refreshWhenEdit = false;
	}

	public SpreadSheetCell(Cell cell) {
		this();
		this.name = cell.getName();
		this.row = cell.getRow();
		this.col = cell.getColumn();
		this.sheetIndex = cell.getSheet();
	}

	@JsonIgnore
	public boolean isReport() {
		return this.type == SpreadSheetType.REPORT;
	}

	@JsonIgnore
	public boolean isInput() {
		return this.type == SpreadSheetType.INPUT;
	}

	@Override
	public String toString() {
		return this.name;
	}

	@Override
	public Persistent copy() {
		return null;
	}

}

File: client/bcephal-client-model/Sheets/SpreadSheetCell.cs
using Bcephal.Models.Base;
using Bcephal.Models.Filters;
using Bcephal.Models.Grids;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Bcephal.Models.Sheets
{
    public class SpreadSheetCell : Persistent
    {

        public SpreadSheetType? Type { get; set; }

        public string Name { get; set; }

        public int Col { get; set; }

        public int Row { get; set; }

        public string SheetName { get; set; }

        public int SheetIndex { get; set; }

        public MeasureFilterItem CellMeasure { get; set; }

        public UniverseFilter Filter { get; set; }

        public bool RefreshWhenEdit { get; set; }


        [JsonIgnore] public string FullName { get { return string.Format("{0}:{1}", this.SheetName, this.Name); } }

        [JsonIgnore]
        public bool IsReport
        {
            get { return this.Type == SpreadSheetType.REPORT; }
        }

        [JsonIgnore]
        public bool IsInput
        {
            get { return this.Type == SpreadSheetType.INPUT; }
        }

        [JsonIgnore]
        public bool IsNew { get; set; }


        public SpreadSheetCell()
        {
            this.RefreshWhenEdit = false;
        }

        public SpreadSheetCell(Cell cell) : this()
        {
            this.Name = cell.Name;
            this.Row = cell.Row;
            this.Col = cell.Column;
            this.SheetIndex = cell.Sheet;
        }


        public override int CompareTo(object obj)
        {
            if (obj == null || !(obj is SpreadSheetCell)) return 1;
            if (this == obj) return 0;
            if (this.Id.HasValue && this.Id.Equals(((SpreadSheetCell)obj).Id)) return 0;
            if (this.SheetIndex.Equals(((SpreadSheetCell)obj).SheetIndex))
            {
                return this.Name.CompareTo(((SpreadSheetCell)obj).Name);
            }
            return this.FullName.CompareTo(((SpreadSheetCell)obj).FullName);
        }

        //public override bool Equals(Object obj)
        //{
        //    return CompareTo(obj) == 0;
        //}

        public override string ToString()
        {
            return this.Name;
        }


    }
}
